File: DS/stacks/stack.js
// Implementation of a stack using an array
let stack = []

// Initial stack
console.log("Initial Stack: ", stack)

// Append to the top of the stack - O(1)
stack.push(1)
stack.push(2)
stack.push(3)
stack.push(4)
stack.push(5)
stack.push(6)

console.log("Stack after appending elements: ", stack)

// Pop from the top of the stack - O(1)
let x = stack.pop()
console.log("Popped element: ", x)

console.log("Stack after popping element: ", stack)

// Peek at the top of the stack - O(1)
let y = stack[stack.length - 1]
console.log("Peeked element: ", y)
console.log("Stack after peeking element: ", stack)

// Check if stack is empty - O(1)
console.log("Is stack empty? ", stack.length === 0)

// Implementing stack using a doubly linked list

File: DS/stacks/stack.cpp
#include <iostream>
#include <stack>

// implemented with <stack>
int main() {
    // Stack - Last in First out

    std::stack<int> stk;
    std::cout << "Initial Stack: empty" << std::endl;

    // append to top of Stack O(1)
    stk.push(5);
    stk.push(4);
    stk.push(3);

    std::cout << "Stack after pushing elements: [5, 4, 3]" << std::endl;

    // Pop from stack top - O(1)

    int x = stk.top();
    stk.pop();

    std::cout << "popped element: " << x << std::endl;
    std::cout << "Stack after popping: [5, 4]" << std::endl;

    // Peeking at the top of the stack - O(1)

    std::cout << "Top element of the stack: " << stk.top() <<  std::endl; 

    // Asking if stack is empty 

    if (stk.empty()) {
        std::cout << "Stack is empty" << std::endl;
    }
}